File: frontend/types/chatbot.ts
/**
 * Chatbot type definitions for arbitrage bot configuration.
 *
 * These types mirror the backend Pydantic models in backend/app/models/chatbot.py
 */

/**
 * Message types in the conversation
 */
export enum MessageType {
  USER = "user",
  ADVISOR = "advisor",
  SYSTEM = "system",
}

/**
 * Steps in the assessment workflow
 */
export enum AssessmentStep {
  ASSESSMENT = "assessment",
  OPTIMIZATION = "optimization",
  TESTING = "testing",
  CONFIGURATION = "configuration",
  ACTIVATION = "activation",
}

/**
 * A single message in the chat conversation
 */
export interface ChatMessage {
  id: string;
  session_id: string;
  role: MessageType;
  content: string;
  timestamp: string; // ISO datetime
  metadata?: Record<string, any>;
}

/**
 * User preferences gathered during assessment
 */
export interface UserPreferences {
  // Core trading parameters
  risk_tolerance?: number; // 1-10
  initial_capital_usd?: number;
  max_trade_size_usd?: number;

  // Experience and goals
  trading_experience?: "beginner" | "intermediate" | "advanced";
  primary_goal?: "max_profit" | "steady_income" | "learning";

  // Market preferences
  preferred_markets?: string[]; // e.g., ["cryptocurrency", "prediction_markets"]
  excluded_markets?: string[];

  // Operational constraints
  monitoring_frequency?: "realtime" | "daily" | "weekly";
  gas_sensitivity?: "low" | "medium" | "high";

  // Risk management
  max_drawdown_tolerance_pct?: number;
  position_sizing_preference?: "conservative" | "moderate" | "aggressive";

  // Additional context
  additional_notes?: string;
  assessment_complete?: boolean;
}

/**
 * An active assessment session
 */
export interface AssessmentSession {
  id: string;
  created_at: string; // ISO datetime
  updated_at: string; // ISO datetime
  current_step: AssessmentStep;
  preferences: UserPreferences;

  // Workflow tracking
  optimizer_session_id?: string;
  is_complete?: boolean;

  // Progress tracking
  required_fields_collected: number;
  total_required_fields: number;
}

/**
 * Request to send a message to the chatbot
 */
export interface ChatRequest {
  session_id: string;
  message: string;
  context_updates?: Record<string, any>;
}

/**
 * Response from the chatbot
 */
export interface ChatResponse {
  session_id: string;
  message_id: string;
  response: string;
  preferences_updated: boolean;
  new_preferences?: UserPreferences;
  assessment_complete: boolean;
  suggested_next_step?: AssessmentStep;
  progress?: string; // e.g., "5/7 fields collected"
}

/**
 * Context information for a session
 */
export interface SessionContext {
  session: AssessmentSession;
  current_parameters?: Record<string, any>;
  baseline_metrics?: Record<string, number>;
  conversation_summary?: string;
}

/**
 * Request to start a new session
 */
export interface StartSessionRequest {
  user_id?: string;
}

/**
 * Response when starting a new session
 */
export interface StartSessionResponse {
  session_id: string;
  initial_message: string;
  current_step: AssessmentStep;
}

/**
 * WebSocket message types
 */
export type WebSocketMessage =
  | { type: "connected"; data: ConnectedData }
  | { type: "advisor_message"; data: AdvisorMessageData }
  | { type: "assessment_complete"; data: AssessmentCompleteData }
  | { type: "context_update"; data: ContextUpdateData }
  | { type: "error"; data: ErrorData }
  | { type: "ping" }
  | { type: "pong" };

export interface ConnectedData {
  session_id: string;
  current_step: AssessmentStep;
  progress: string;
}

export interface AdvisorMessageData {
  response: string;
  preferences_updated: boolean;
  new_preferences?: UserPreferences;
  assessment_complete: boolean;
  progress?: string;
}

export interface AssessmentCompleteData {
  session_id: string;
  preferences: UserPreferences;
  next_step: string;
}

export interface ContextUpdateData {
  preferences: UserPreferences;
  conversation_history: ChatMessage[];
  progress: string;
}

export interface ErrorData {
  message: string;
}

/**
 * Quick action button configuration
 */
export interface QuickAction {
  label: string;
  value: string;
  category: "risk" | "experience" | "goal" | "market" | "general";
}

/**
 * Step panel props
 */
export interface StepPanelProps {
  sessionId: string;
  onNextStep: () => void;
  onPreviousStep: () => void;
  currentStep: AssessmentStep;
}

/**
 * Workflow state
 */
export interface WorkflowState {
  currentStep: AssessmentStep;
  sessionId: string | null;
  preferences: UserPreferences | null;
  optimizerSessionId?: string;
  isLoading: boolean;
  error?: string;
}

File: backend/app/models/chatbot.py
"""
Chatbot data models for arbitrage bot configuration.

Defines schemas for assessment sessions, messages, and user preferences.
"""

from datetime import datetime
from typing import Optional, Dict, Any, List
from enum import Enum
from pydantic import BaseModel, Field


class MessageType(str, Enum):
    """Types of messages in the conversation."""
    USER = "user"
    ADVISOR = "advisor"
    SYSTEM = "system"


class AssessmentStep(str, Enum):
    """Steps in the assessment workflow."""
    ASSESSMENT = "assessment"           # Step 1: Chat conversation
    OPTIMIZATION = "optimization"       # Step 2: Ring consensus
    TESTING = "testing"                 # Step 3: Monte Carlo simulation
    CONFIGURATION = "configuration"     # Step 4: Final approval
    ACTIVATION = "activation"           # Step 5: Bot is live


class ChatMessage(BaseModel):
    """A single message in the chat conversation."""
    id: str = Field(..., description="Unique message identifier")
    session_id: str = Field(..., description="Session this message belongs to")
    role: MessageType = Field(..., description="Who sent the message")
    content: str = Field(..., description="Message content")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="When message was sent")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional message data")


class UserPreferences(BaseModel):
    """User preferences gathered during assessment."""
    # Core trading parameters
    risk_tolerance: Optional[int] = Field(None, ge=1, le=10, description="Risk tolerance 1-10")
    initial_capital_usd: Optional[float] = Field(None, gt=0, description="Initial trading capital")
    max_trade_size_usd: Optional[float] = Field(None, gt=0, description="Maximum per-trade size")

    # Experience and goals
    trading_experience: Optional[str] = Field(None, description="Experience level: beginner/intermediate/advanced")
    primary_goal: Optional[str] = Field(None, description="Main goal: max_profit/steady_income/learning")

    # Market preferences
    preferred_markets: Optional[List[str]] = Field(default_factory=list, description="Markets to trade")
    excluded_markets: Optional[List[str]] = Field(default_factory=list, description="Markets to avoid")

    # Operational constraints
    monitoring_frequency: Optional[str] = Field(None, description="How often user checks: realtime/daily/weekly")
    gas_sensitivity: Optional[str] = Field(None, description="Gas cost concern: low/medium/high")

    # Risk management
    max_drawdown_tolerance_pct: Optional[float] = Field(None, ge=0, le=100, description="Max acceptable drawdown %")
    position_sizing_preference: Optional[str] = Field(None, description="Position sizing: conservative/moderate/aggressive")

    # Additional context
    additional_notes: Optional[str] = Field(None, description="Any extra information provided")
    assessment_complete: bool = Field(default=False, description="Whether assessment is complete")


class AssessmentSession(BaseModel):
    """An active assessment session."""
    id: str = Field(..., description="Unique session identifier")
    created_at: datetime = Field(default_factory=datetime.utcnow, description="Session start time")
    updated_at: datetime = Field(default_factory=datetime.utcnow, description="Last update time")
    current_step: AssessmentStep = Field(default=AssessmentStep.ASSESSMENT, description="Current workflow step")
    preferences: UserPreferences = Field(default_factory=UserPreferences, description="Gathered preferences")

    # Workflow tracking
    optimizer_session_id: Optional[str] = Field(None, description="Linked optimizer session")
    is_complete: bool = Field(default=False, description="Whether entire workflow is complete")

    # Progress tracking
    required_fields_collected: int = Field(default=0, description="Number of required fields gathered")
    total_required_fields: int = Field(default=7, description="Total required fields to collect")


class ChatRequest(BaseModel):
    """Request to send a message to the chatbot."""
    session_id: str = Field(..., description="Session identifier")
    message: str = Field(..., description="User's message content")
    context_updates: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Any context updates")


class ChatResponse(BaseModel):
    """Response from the chatbot."""
    session_id: str = Field(..., description="Session identifier")
    message_id: str = Field(..., description="Response message identifier")
    response: str = Field(..., description="Advisor's response text")
    preferences_updated: bool = Field(default=False, description="Whether preferences were updated")
    new_preferences: Optional[UserPreferences] = Field(None, description="Updated preferences")
    assessment_complete: bool = Field(default=False, description="Whether assessment is now complete")
    suggested_next_step: Optional[AssessmentStep] = Field(None, description="Suggested next workflow step")
    progress: Optional[str] = Field(None, description="Progress indicator (e.g., '5/7 fields collected')")


class SessionContext(BaseModel):
    """Context information for a session."""
    session: AssessmentSession = Field(..., description="The session data")
    current_parameters: Optional[Dict[str, Any]] = Field(None, description="Current bot parameters")
    baseline_metrics: Optional[Dict[str, float]] = Field(None, description="Current performance metrics")
    conversation_summary: Optional[str] = Field(None, description="Summary of conversation so far")


class StartSessionRequest(BaseModel):
    """Request to start a new assessment session."""
    user_id: Optional[str] = Field(None, description="Optional user identifier")


class StartSessionResponse(BaseModel):
    """Response when starting a new session."""
    session_id: str = Field(..., description="New session identifier")
    initial_message: str = Field(..., description="First advisor message")
    current_step: AssessmentStep = Field(..., description="Current workflow step")


class TransitionRequest(BaseModel):
    """Request to transition to next step."""
    session_id: str = Field(..., description="Session identifier")
    target_step: AssessmentStep = Field(..., description="Step to transition to")


class TransitionResponse(BaseModel):
    """Response for step transition."""
    success: bool = Field(..., description="Whether transition succeeded")
    current_step: AssessmentStep = Field(..., description="New current step")
    message: str = Field(..., description="Transition message")
    data: Optional[Dict[str, Any]] = Field(None, description="Additional data for the step")
